File: runtime/go/prompty/jinjasubset/ast.go
package jinjasubset

type expr interface{ exprNode() }

type litExpr struct{ value any }
type varExpr struct {
	root string
	path []pathSeg
}
type filterExpr struct {
	name  string
	input expr
	args  []expr
}
type unaryExpr struct {
	op      string
	operand expr
}
type binaryExpr struct {
	op          string
	left, right expr
}

func (litExpr) exprNode()    {}
func (varExpr) exprNode()    {}
func (filterExpr) exprNode() {}
func (unaryExpr) exprNode()  {}
func (binaryExpr) exprNode() {}

type pathSeg interface{ pathNode() }

type attrSeg struct{ name string }
type indexSeg struct{ expr expr }

func (attrSeg) pathNode()  {}
func (indexSeg) pathNode() {}

type node interface{ nodeMarker() }

type textNode struct{ value string }
type interpNode struct{ expr expr }
type branch struct {
	test expr
	body []node
}
type ifNode struct {
	branches []branch
	elseBody []node
}
type forNode struct {
	loopVar string
	seq     expr
	body    []node
}

func (textNode) nodeMarker()   {}
func (interpNode) nodeMarker() {}
func (ifNode) nodeMarker()     {}
func (forNode) nodeMarker()    {}

File: runtime/typescript/packages/core/src/jinja-subset/ast.ts
export type Expr =
  | { kind: "lit"; value: unknown }
  | { kind: "var"; root: string; path: PathSeg[] }
  | { kind: "filter"; name: string; input: Expr; args: Expr[] }
  | { kind: "unary"; operator: "not"; operand: Expr }
  | { kind: "binary"; operator: string; left: Expr; right: Expr };

export type PathSeg =
  { kind: "attr"; name: string } | { kind: "index"; expr: Expr };

export type Node =
  | { kind: "text"; value: string }
  | { kind: "interp"; expr: Expr }
  | { kind: "if"; branches: Branch[]; elseBody?: Node[] }
  | { kind: "for"; loopVar: string; seq: Expr; body: Node[] };

export interface Branch {
  test: Expr;
  body: Node[];
}
